### DISCUSSION TITLE
[init_implicit]How to enable a tuple to be implicit converted to a class?

### DISCUSSION THREAD

**Discussion by huangweiwu** [wjakob/nanobind#290] [Q&A] [answered]
I have a class point below, but the init_implicit code comes up with a compile error.
Cannot convert from 'initializer list' to 'point'
```
nb::class_<point>(m, "point)
    .def(nb::init<float, float>())
    .def(nb::init_implicit<std::tuple<float, float>>())

**Comment by wjakob** [comment]
Do you have a constructor taking that exact argument? Perhaps you are missing a `&` or `const`

**Reply by huangweiwu** [reply]
I do have that constructor taking (const float&, const float&), but it still comes up with the same error.
```
#include "nanobind/stl/tuple.h"
nb::class_<point>(m, "point")
    .def(nb::init<const float&, const float&>())
    .def(nb::init_implicit<std::tuple<const float&, const float&>>())

**Reply by wjakob** [reply]
Then you will need to post a small reproducer.

**Reply by huangweiwu** [reply]
I add a constructor taking(std::tuple<const float&, const float&>) and it works.
But I have to bind both float and int so as to enable passing int value from python.
Int is not permitted to implicit transform to float in this case?
```
#include "nanobind/stl/tuple.h"
nb::class_<point>(m, "point")
    .def(nb::init<const float&, const float&>())
    .def(nb::init_implicit<std::tuple<const float&, const float&>>())
    .def(nb::init_implicit<std::tuple<const int&, const int&>>())

**Reply by wjakob** [reply]
It's really difficult to comment without having a reproducer: complete binding code, and the way you call it form Python.

**Reply by huangweiwu** [reply]
The complete binding code
```
#include "nanobind/stl/tuple.h"

template <typename T>
struct point {
    T x;
    T y;
    point() {}
    point(const T& _x,const T& _y) : x(_x), y(_y) {}
    point(std::tuple<const T&, const T&> _t) : x(std::get<0>(_t)), y(std::get<1>(_t)) {}
}

def test_func(point p)
{
}

nb::class_<point>(m, "point")
    .def(nb::init<const float&, const float&>())
    .def(nb::init_implicit<std::tuple<const float&, const float&>>())
    .def(nb::init_implicit<std::tuple<const int&, const int&>>())
m.def(test_func, "test_func")
```
Call it from python.
```
test_func((1.0, 1.0))
test_func((1, 1))
```

**Reply by wjakob** [reply]
Aha. Yes, that of does not work because this constructor does not exist, and implicit conversion from float to int won't be done. The documentation explains how you can make a custom `__init__` style constructor and sign it up for implicit conversions.

**Reply by huangweiwu** [reply]
Boost python has a way to register default implicit conversion.
Using boost::python::converter::registry to register check function and construct function.
In nanobind, is there any other ways to make default implicit conversion besides init_implicit.

**Reply by wjakob** [reply]
Not really an officially supported way, but you can do:

```cpp
nb::detail::implicitly_convertible(
    [](PyTypeObject *dst, PyObject *src, detail::cleanup_list *cleanup) noexcept -> bool {
        /* Return whether or not `src` is implicitly convertible to a Python type `dst`. */
    },
    &typeid(TargetCPPType) /* <-- this is the C++ type associated with the binding */);
```

This is what nanobind uses to decide whether or not to call a constructor of some class. That constructor must actually exist though.

**Reply by huangweiwu** [reply, marked as answer]
It seems that the corresponding constructor is indispensable.
